src/viem/saga.test.ts

```diff
@@ -0,0 +1,220 @@
+      ])
+      .call(getViemWallet, networkConfig.viemChain.celo)
+      .not.call.fn(encryptComment)
+      .call(getSendTxFeeDetails, {
+        recipientAddress: mockSendPaymentArgs.recipientAddress,
+        amount: BigNumber(2),
+        tokenAddress: mockCeloAddress,
+        feeInfo: mockFeeInfo,
+      })
+      .returns(mockTxHash)
+      .run()
+
+    expect(simulateContractSpy).toHaveBeenCalledWith({
+      address: getAddress(mockCeloAddress),
+      abi: erc20.abi,
+      functionName: 'transfer',
+      account: mockViemWallet.account,
+      args: [getAddress(mockSendPaymentArgs.recipientAddress), BigInt(2e18)],
+      ...mockViemFeeInfo,
+    })
+  })
+
+  it('throws if simulateContract fails', async () => {
+    simulateContractSpy.mockRejectedValue(new Error('simulate error'))
+
+    await expectSaga(sendPayment, { ...mockSendPaymentArgs, tokenAddress: mockCeloAddress })
+      .withState(createMockStore().getState())
+      .provide([
+        [matchers.call.fn(getViemWallet), mockViemWallet],
+        [matchers.call.fn(getSendTxFeeDetails), mockViemFeeInfo],
+      ])
+      .throws(new Error('simulate error'))
+      .run()
+  })
+
+  it('throws if writeContract fails', async () => {
+    mockViemWallet.writeContract.mockRejectedValue(new Error('write error'))
+
+    await expectSaga(sendPayment, { ...mockSendPaymentArgs, tokenAddress: mockCeloAddress })
+      .withState(createMockStore().getState())
+      .provide([
+        [matchers.call.fn(getViemWallet), mockViemWallet],
+        [matchers.call.fn(getSendTxFeeDetails), mockViemFeeInfo],
+      ])
+      .throws(new Error('write error'))
+      .run()
+  })
+})
+
+describe('getSendTxFeeDetails', () => {
+  it('calls buildSendTx and chooseTxFeeDetails with the expected values', async () => {
+    const recipientAddress = mockAccount
+    const amount = new BigNumber(10)
+    const tokenAddress = mockCusdAddress
+    const feeInfo = mockFeeInfo
+    const celoTx = {
+      txo: 'test',
+    } as unknown as CeloTransactionObject<unknown>
+    const encryptedComment = 'test'
+
+    const mockFeeDetails = {
+      feeCurrency: mockCusdAddress,
+      gas: feeInfo.gas,
+      gasPrice: feeInfo.gasPrice,
+    }
+
+    await expectSaga(getSendTxFeeDetails, {
+      recipientAddress,
+      amount,
+      tokenAddress,
+      feeInfo,
+      encryptedComment,
+    })
+      .withState(createMockStore().getState())
+      .provide([
+        [call(buildSendTx, tokenAddress, amount, recipientAddress, encryptedComment), celoTx],
+        [
+          call(
+            chooseTxFeeDetails,
+            celoTx.txo,
+            feeInfo.feeCurrency,
+            feeInfo.gas.toNumber(),
+            feeInfo.gasPrice
+          ),
+          mockFeeDetails,
+        ],
+      ])
+      .returns(mockViemFeeInfo)
+      .run()
+  })
+
+  // TODO(any): remove this test once we've ensured gas and gasPrice are always BigNumbers
```

Comment:
It seems like it might also be appropriate to put a similar comment by chooseTxFeeDetails.

And since this same theme of ensuring fee types will have TODO's in three places, can you make a ticket just to keep all the details in one place?

Reply:
how do you suggest that we ensure gas/gasPrice are always BigNumber?

Reply:
the todo is labeled "any", but I don't think anyone could pick this up-- as a non-implementer, I don't really know what's meant to be done here or how I would do it. or why it's desirable

Reply:
Some additional context - we cache fees in Redux per token. When we write the fees to Redux, they're stored as BigNumbers. If there are no fees already present in the cache when we're fetching fees, we calculate a new fee and store it Redux; when we do this, any code that accesses this new fee shortly after it's written to Redux will access it as a BigNumber, however, on _subsequent_ accesses later on, the fee will be returned from Redux as a _string_. I think this has something to do with when/how Redux asynchronously serializes values to the store (we store the actual BN object, but Redux serializes it to a string when the store is actually updated) - some kind of race condition.

One potential solution to this is to just update the fee selector to deserialize the value in Redux (string or otherwise) to a BN object, though that's kind of a workaround to the bad practice of us _not_ serializing the BN before writing it to Redux in the first place.

Reply:
I'll make an issue for this.